File: 189-RotateArray/189-RotateArray.java
// Last updated: 16/10/2025, 04:28:19
class Solution { 

    public void reverse(int [] nums, int left, int right, int n){
        while(left < right){
            int temp = nums[left];
            nums[left++] = nums[right];
            nums[right--] = temp;
        }
    }

    public void rotate(int[] nums, int k) {
        int n = nums.length;
        if(n == 0 || k % n == 0) return;
        k = k%n;

        reverse(nums, 0, n - k - 1, n);
        reverse(nums, n - k, n - 1, n);
        reverse(nums,0, n - 1, n);  
        
    }
}

//Note: The subList method in Java returns a view of the original list, and changes made to the view are reflected in the original list.

      

File: 189-RotateArray/189-RotateArray.cpp
// Last updated: 16/10/2025, 04:28:25
class Solution {
public:
    void reverse(vector<int> &nums, int s, int e){
        int i = s;
        int j = e;
        while(i < j){
            int temp = nums[i];
            nums[i] = nums[j];
            nums[j] = temp;
            i++;
            j--;
        }
    }
    
    void rotate(vector<int>& nums, int k) {
        int n = nums.size();
        if(n == 0 || n == 1){
            return;
        }
        k = k%n;
        reverse(nums,0,n-1);
        reverse(nums,0,k-1);
        reverse(nums,k,n-1);
    }
};